### PR TITLE
#122 Added 5 New Algorithms under String Algorithms

### LINKED ISSUE #122: [Feature Request] Addition of Apostolico–Giancarlo Algorithm,BNDM , Bitap Algorithm , Commentz-Walter Algorithm, Shift-Or Algorithm in String Algorithms
### Describe the solution you'd like

Apostolico–Giancarlo Algorithm
Description: An extension of the Boyer-Moore algorithm optimized for multiple pattern matching, allowing efficient backtracking.
Time Complexity: O(n + m)

BNDM (Backward Nondeterministic Dawg Matching)
Description: A bitwise pattern matching algorithm that is an improvement over the BMH algorithm, especially effective for short patterns.
Time Complexity: O(nm)

Bitap Algorithm (Approximate String Matching)
Description: Uses bitwise operations to perform approximate pattern matching, allowing a specified number of errors (insertions, deletions, or substitutions).
Time Complexity: O(nm)

Commentz-Walter Algorithm
Description: A variation of the Boyer-Moore algorithm that improves efficiency when searching for multiple patterns simultaneously.
Time Complexity: O(n + m)

Shift-Or Algorithm
Description: A bitwise algorithm that matches patterns by shifting a bitmask, making it effective for small alphabets.
Time Complexity: O(n)

**Would you like to work on this feature?**  
- [✓ ] Yes  
- [ ] No  

@ajay-dhangar  Please assign me this issue.I would like to work on it.

### PR DESCRIPTION
**Pull request by pratheekv39**
### Description

1. Summary of Change: Implemented 5 algorithms under the String Algorithms section: 
   - Apostolico–Giancarlo Algorithm
   - BNDM (Backward Nondeterministic Dawg Matching) Algorithm
   - Bitap Algorithm (Approximate String Matching)
   - Commentz-Walter Algorithm
   - Shift-Or Algorithm

2. Issue Fixed: Resolved issue #122  by adding the necessary algorithms to the string algorithms section.

3. Motivation: The changes enhance the algorithm coverage and provide detailed implementations for each string matching algorithm.

4. Context: Detailed comments were added where necessary to improve clarity and code understanding.

5. Dependencies:  No new dependencies introduced.

Fixes #122 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [ X] My code follows the style guidelines of this project
- [ X] I have performed a self-review of my code
- [ X] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [X ] New and existing unit tests pass locally with my changes
- [X ] Any dependent changes have been merged and published in downstream modules

@ajay-dhangar  Please add the required labels GSSOC Ext and Hacktoberfest . Thanks....

### CODE REVIEW

**Comment by pratheekv39** [comment]
@ajay-dhangar Can you please review it and add GSSOC Ext and Hacktoberfest labels to it . Thanks......

File: docs/algorithms/string-algorithms/ApostolicoGiancarlo.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+- **Suboptimal on Small Patterns**:
+  - While efficient for longer patterns, its performance may not be as significant for smaller ones compared to simpler algorithms like Knuth-Morris-Pratt (KMP).
+
+### Time Complexity:
+
+- **Best Case: O(n/m)**  
+  In the best-case scenario, the algorithm performs optimally, making only a fraction of comparisons proportional to the length of the text divided by the length of the pattern.
+
+- **Average Case: O(n)**
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/ApostolicoGiancarlo.md]: replace `O(n)` with `$O(n)$`

File: docs/algorithms/string-algorithms/ApostolicoGiancarlo.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+### Time Complexity:
+
+- **Best Case: O(n/m)**  
+  In the best-case scenario, the algorithm performs optimally, making only a fraction of comparisons proportional to the length of the text divided by the length of the pattern.
+
+- **Average Case: O(n)**  
+  On average, the Apostolico–Giancarlo algorithm makes approximately linear scans through the text, resulting in efficient performance for most practical use cases.
+
+- **Worst Case: O(n * m)**
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/ApostolicoGiancarlo.md]: replace `O(n * m)` with `$O(n \times m)$`

File: docs/algorithms/string-algorithms/ApostolicoGiancarlo.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+- **Average Case: O(n)**  
+  On average, the Apostolico–Giancarlo algorithm makes approximately linear scans through the text, resulting in efficient performance for most practical use cases.
+
+- **Worst Case: O(n * m)**  
+  In the worst case, if the pattern has repeated sections that align poorly with the text, the algorithm could degrade to quadratic time complexity, where `n` is the text length and `m` is the pattern length.
+
+### Space Complexity:
+
+- **Space Complexity: O(m + n)**
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/ApostolicoGiancarlo.md]: replace `O(m + n)` with `$O(m + n)$`

File: docs/algorithms/string-algorithms/BNDM.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+- **Efficient for Short Patterns**:
+  - BNDM is particularly efficient for short patterns, often outperforming other string matching algorithms like Boyer-Moore and Knuth-Morris-Pratt for small pattern sizes.
+
+- **Extension of BDM**:
+  - It improves upon the BDM algorithm by handling more general cases and providing better performance for non-trivial patterns.
+
+### Time Complexity:
+
+- **Best Case: O(n / w)**
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/BNDM.md]: replace `O(n / w)` with `$O\left(\frac{n}{w}\right)$`

File: docs/algorithms/string-algorithms/BNDM.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+- **Extension of BDM**:
+  - It improves upon the BDM algorithm by handling more general cases and providing better performance for non-trivial patterns.
+
+### Time Complexity:
+
+- **Best Case: O(n / w)**  
+  In the best-case scenario, where `w` is the word size of the machine, the algorithm takes advantage of the word-level parallelism and makes few character comparisons.
+
+- **Average Case: O(n)**
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/BNDM.md]: replace `O(n)` with `$O(n)$`

File: docs/algorithms/string-algorithms/BNDM.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+### Time Complexity:
+
+- **Best Case: O(n / w)**  
+  In the best-case scenario, where `w` is the word size of the machine, the algorithm takes advantage of the word-level parallelism and makes few character comparisons.
+
+- **Average Case: O(n)**  
+  On average, BNDM performs linear scans through the text, making it highly efficient for typical use cases, especially with short patterns.
+
+- **Worst Case: O(n * m)**
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/BNDM.md]: replace `O(n * m)` with `$O(n \times m)$`

File: docs/algorithms/string-algorithms/BNDM.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+- **Average Case: O(n)**  
+  On average, BNDM performs linear scans through the text, making it highly efficient for typical use cases, especially with short patterns.
+
+- **Worst Case: O(n * m)**  
+  In the worst case, when the text and pattern have poor alignment, BNDM may require multiple full scans of the text, leading to quadratic complexity, where `n` is the text length and `m` is the pattern length.
+
+### Space Complexity:
+
+- **Space Complexity: O(m)**
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/BNDM.md]: replace `O(m)` with `$O(m)$`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
Hi @pratheekv39,

Thank you for your contributions! I have a few suggestions to improve the consistency and readability of our project:

1. **File Naming Best Practices**: Please rename the following files to follow a consistent, clear naming convention:
   - `ApostolicoGiancarlo.md` → `apostolico-giancarlo.md`
   - `BNDM.md` → `bndm.md`
   - `Bitap.md` → `bitap.md`
   - `CommentzWalter.md` → `commentz-walter.md`
   - `ShiftOr.md` → `shift-or.md`

   Using lowercase letters and hyphens between words makes file names more uniform and easier to manage across different platforms.

2. **LaTeX Syntax in Markdown Files**: When writing mathematical expressions in LaTeX, please ensure that you're using the correct syntax. For inline math, use single dollar signs (`$ ... $`), and for block math, use double dollar signs (`$$ ... $$`). This ensures proper rendering across all markdown viewers that support LaTeX.

Example for inline math: `$ O\left(\frac{n}{w}\right) $`

Example for block math:
```
$$
O\left(\frac{n}{w}\right)
$$
```

Let me know if you need any assistance with these changes.

Best regards,  
@ajay-dhangar

**Comment by pratheekv39** [comment]
@ajay-dhangar I have made the changes that you requested . Please check once...

**Review by ajay-dhangar — APPROVED** [review]
Done

**Comment by ajay-dhangar** [comment]
> @ajay-dhangar I have made the changes that you requested . Please check once...

@pratheekv39, you need to sync our latest code. I've resolved the error, so first, sync with `git pull origin main`, push the code to your branch, and then resolve any remaining errors.

**Comment by pratheekv39** [comment]
@ajay-dhangar Ok sir , I will sync the lastest code and send the updates by Monday . Been busy with some work lately:(

**Comment by pratheekv39** [comment]
@ajay-dhangar I have resolved and updated (synced) my branch and pushed the code. Can you now approve the PR???

**Pull request merged**